### PR TITLE
Literature restful

### PR DESCRIPTION
**Pull request by LeonardoGonzales**

### CODE REVIEW

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/output/LiteratureEntryFilter.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import uk.ac.ebi.uniprot.search.field.LiteratureField;
+
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * @author lgonzales
+ * @since 2019-07-04
+ */
+public class LiteratureEntryFilter {
```
Comment (sahmad-ebi) [inline support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/output/LiteratureEntryFilter.java]: Need better approach to unset the fields with null. I am looking into it.

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/service/LiteratureService.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    public Stream<LiteratureEntry> download(LiteratureRequestDTO request) {
+        SolrRequest solrRequest = basicService.createSolrRequest(request, facetConfig, literatureSortClause, defaultSearchHandler);
+        return basicService.download(solrRequest);
+    }
+
+    public QueryResult<LiteratureEntry> getMappedLiteratureByUniprotAccession(final String accession, LiteratureMappedRequestDTO requestDTO) {
+        SolrRequest solrRequest = SolrRequest.builder()
+                .query("mapped_protein:" + accession)
+                .addSort(Sort.by(Sort.Direction.ASC, LiteratureField.Search.id.name()))
```
Comment (sahmad-ebi) [inline support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/service/LiteratureService.java]: Should we not put sort field in LiteratureMappedRequestDTO and add default asc on id like i have done in disease service

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/support_data/config/MessageConverterConfig.java [hunk 3]
```diff
@@ -47,6 +50,10 @@ public void extendMessageConverters(List<HttpMessageConverter<?>> converters) {
                 converters.add(new ErrorMessageConverter());
                 converters.add(new ListMessageConverter());
 
+                converters.add(new LiteratureXlsMessageConverter());
+                converters.add(new LiteratureTsvMessageConverter());
+                converters.add(0, new LiteratureJsonMessageConverter());
```
Comment (sahmad-ebi) [inline support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/support_data/config/MessageConverterConfig.java]: please document index 0 for json.

Comment (LeonardoGonzales) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/support_data/config/MessageConverterConfig.java]: improved and code documented

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/service/LiteratureService.java (hunk 2 above)
Comment (LeonardoGonzales) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/service/LiteratureService.java]: Excellent catch!!

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/output/LiteratureEntryFilter.java (hunk 1 above)
Comment (LeonardoGonzales) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/output/LiteratureEntryFilter.java]: ok

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/service/LiteratureService.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                .build();
+        QueryResult<LiteratureDocument> results = repository.searchPage(solrRequest, requestDTO.getCursor(), requestDTO.getSize());
+
+        List<LiteratureEntry> converted = results.getContent().stream()
+                .map(new LiteratureEntryConverter())
+                .filter(Objects::nonNull)
+                .peek(literatureEntry -> {
```
[the remark this answers is not in the document]
Comment (LeonardoGonzales) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/service/LiteratureService.java]: I improved it. please, let me know if it is good enough

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/service/LiteratureService.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    public QueryResult<LiteratureEntry> getMappedLiteratureByUniprotAccession(final String accession, LiteratureMappedRequestDTO requestDTO) {
+        SolrRequest solrRequest = SolrRequest.builder()
+                .query("mapped_protein:" + accession)
+                .addSort(Sort.by(Sort.Direction.ASC, LiteratureField.Search.id.name()))
+                .build();
+        QueryResult<LiteratureDocument> results = repository.searchPage(solrRequest, requestDTO.getCursor(), requestDTO.getSize());
+
+        List<LiteratureEntry> converted = results.getContent().stream()
+                .map(new LiteratureEntryConverter())
```
[the remark this answers is not in the document]
Comment (LeonardoGonzales) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/service/LiteratureService.java]: I improved it, please, let me know if it is good enough

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/output/LiteratureEntryFilter.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+    private static String DEFAULT_FIELDS = "id,title,reference,lit_abstract";
```
[the remark this answers is not in the document]
Comment (LeonardoGonzales) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/output/LiteratureEntryFilter.java]: I think you are working on it, right?

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/LiteratureController.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+                                                        @RequestHeader(value = "Accept", defaultValue = APPLICATION_JSON_VALUE)
+                                                                MediaType contentType,
+                                                        @RequestHeader(value = "Accept-Encoding", required = false)
+                                                                String encoding,
+                                                        HttpServletRequest request) {
+        Stream<LiteratureEntry> result = literatureService.download(searchRequest);
+        return super.download(result, searchRequest.getFields(), contentType, request, encoding);
+    }
+
+    @GetMapping(value = "/mapped/proteins/{accession}", produces = {TSV_MEDIA_TYPE_VALUE, LIST_MEDIA_TYPE_VALUE, APPLICATION_JSON_VALUE, XLS_MEDIA_TYPE_VALUE})
```
[the remark this answers is not in the document]
Comment (LeonardoGonzales) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/LiteratureController.java]: I did not understand what your point. can you please show me an example please?

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/LiteratureController.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    @RequestMapping(value = "/download", method = RequestMethod.GET,
+            produces = {TSV_MEDIA_TYPE_VALUE, LIST_MEDIA_TYPE_VALUE, APPLICATION_JSON_VALUE, XLS_MEDIA_TYPE_VALUE})
+    public ResponseEntity<ResponseBodyEmitter> download(@Valid
+                                                                LiteratureRequestDTO searchRequest,
+                                                        @RequestHeader(value = "Accept", defaultValue = APPLICATION_JSON_VALUE)
+                                                                MediaType contentType,
+                                                        @RequestHeader(value = "Accept-Encoding", required = false)
+                                                                String encoding,
+                                                        HttpServletRequest request) {
+        Stream<LiteratureEntry> result = literatureService.download(searchRequest);
```
[the remark this answers is not in the document]
Comment (LeonardoGonzales) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/LiteratureController.java]: I am happy without this.

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/LiteratureController.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+    @RequestMapping(value = "/search", method = RequestMethod.GET,
+            produces = {TSV_MEDIA_TYPE_VALUE, LIST_MEDIA_TYPE_VALUE, APPLICATION_JSON_VALUE, XLS_MEDIA_TYPE_VALUE})
+    public ResponseEntity<MessageConverterContext<LiteratureEntry>> search(@Valid
+                                                                                   LiteratureRequestDTO searchRequest,
+                                                                           @RequestHeader(value = "Accept", defaultValue = APPLICATION_JSON_VALUE)
+                                                                                   MediaType contentType,
+                                                                           HttpServletRequest request,
+                                                                           HttpServletResponse response) {
+        String updatedFields = updateFieldsWithoutMappedReferences(searchRequest.getFields(), contentType);
```
[the remark this answers is not in the document]
Comment (LeonardoGonzales) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/LiteratureController.java]: Good question, I have not found a better way

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/LiteratureController.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+    public LiteratureController(ApplicationEventPublisher eventPublisher, LiteratureService literatureService,
+                                @Qualifier("literatureMessageConverterContextFactory") MessageConverterContextFactory<LiteratureEntry> literatureMessageConverterContextFactory,
+                                ThreadPoolTaskExecutor downloadTaskExecutor) {
+        super(eventPublisher, literatureMessageConverterContextFactory, downloadTaskExecutor, LITERATURE);
+        this.literatureService = literatureService;
+    }
+
+    @GetMapping(value = "/{literatureId}", produces = {TSV_MEDIA_TYPE_VALUE, LIST_MEDIA_TYPE_VALUE, APPLICATION_JSON_VALUE, XLS_MEDIA_TYPE_VALUE})
+    public ResponseEntity<MessageConverterContext<LiteratureEntry>> getById(@PathVariable("literatureId")
```
[the remark this answers is not in the document]
Comment (LeonardoGonzales) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/literature/LiteratureController.java]: Code improved

**Pull request merged**